packages/gatsby-source-contentful/src/types.js

```diff
@@ -0,0 +1,406 @@
+const _ = require(`lodash`)
+
+const typePrefix = `Contentful`
+const makeTypeName = type => _.upperFirst(_.camelCase(`${typePrefix} ${type}`))
+
+exports.makeTypeName = makeTypeName
+
+const makeTextNodeType = (contentTypeName, key) =>
+  _.camelCase(`${contentTypeName} ${key} TextNode`)
+
+exports.makeTextNodeType = makeTextNodeType
+
+const fieldTypeToGraphQLTypeLookup = {
+  // Short string
+  Symbol: () => `String`,
+
+  // Multiple values - handle inner type as usual and just
+  // wrap type in `[]`
+  Array: (field, context) => {
+    const innerField = fieldTypeToGraphQLType(field.items, {
+      isArray: true,
+      ...context,
+    })
+    if (innerField) {
+      return {
+        ...innerField,
+        type: `[${innerField.type}]`,
+      }
+    }
+    return null
+  },
+
+  // Long string (markdown)
+  Text: (field, context) => {
+    const textTypeName = makeTextNodeType(
+      makeTypeName(context.typeName),
+      context.fieldName
+    )
+
+    const textGQLTypeConfig = context.schema.buildObjectType({
+      name: textTypeName,
+      fields: {
+        [context.fieldName]: `String`,
+        // missing - how to properly add childMarkdownRemark field (and other children that
```

Comment:
So children stuff happens outside normal inference flow, so if you add Remark children to Text types when creating them, it should work.

Reply:
OK, disregard that.

I think we can add extension "potentialChildren" that allows specifying which children fields to create.

Reply:
However, that doesn't solve children actually appearing inside the type.

Reply:
This is unsolved problem - not quite sure if it's blocker or not - let my edit PR description. 

My current thinking is that ideally I could add metadata that this node type represents/contains `text/markdown` and transformer plugins could subscribe to that and attach children node type to it.

I don't think we should really specify that children of this type will be specifically `RemarkMarkdown` (we would also need to specify `Mdx` as well) - it's not source plugin job to know what transformers are able to handle their nodes (and we will potentially miss 3rd party or custom transformer plugins this way)

Reply:
Update here for anyone interested. We will ship this without solving children/parent/transformer + schema customization completely, but will document how to workaround this (as temporary solution). The workaround depends on https://github.com/gatsbyjs/gatsby/pull/14116 (being able to define most of the type inside contentful plugin and allowing users to add `childMarkdownRemark` or `childMdx` in project's gatsby-node - and that part will be documented)